homeassistant/helpers/script.py

```diff
@@ -47,107 +49,126 @@ def __init__(self, hass: HomeAssistant, sequence, name: str=None,
+        self._cur = -1
+        self.last_action = None
+        self._trigger_change_listener()
 
     def stop(self) -> None:
         """Stop running script."""
-        with self._lock:
-            if self._cur == -1:
-                return
+        run_callback_threadsafe(self.hass.loop, self.async_stop).result()
 
-            self._cur = -1
-            self._remove_listener()
-            if self._change_listener:
-                self._change_listener()
+    def async_stop(self) -> None:
+        """Stop running script."""
+        if self._cur == -1:
+            return
 
-    def _call_service(self, action, variables):
+        self._cur = -1
+        self._async_remove_listener()
+        self._trigger_change_listener()
+
+    @asyncio.coroutine
+    def _async_call_service(self, action, variables):
         """Call the service specified in the action."""
         self.last_action = action.get(CONF_ALIAS, 'call service')
         self._log("Executing step %s" % self.last_action)
-        service.call_from_config(self.hass, action, True, variables,
-                                 validate_config=False)
+        yield from service.async_call_from_config(
+            self.hass, action, True, variables, validate_config=False)
 
-    def _fire_event(self, action):
+    def _async_fire_event(self, action):
         """Fire an event."""
         self.last_action = action.get(CONF_ALIAS, action[CONF_EVENT])
         self._log("Executing step %s" % self.last_action)
-        self.hass.bus.fire(action[CONF_EVENT], action.get(CONF_EVENT_DATA))
+        self.hass.bus.async_fire(action[CONF_EVENT],
+                                 action.get(CONF_EVENT_DATA))
 
-    def _check_condition(self, action, variables):
+    def _async_check_condition(self, action, variables):
         """Test if condition is matching."""
         self.last_action = action.get(CONF_ALIAS, action[CONF_CONDITION])
-        check = condition.from_config(action, False)(self.hass, variables)
+        check = condition.async_from_config(action, False)(
+            self.hass, variables)
         self._log("Test condition {}: {}".format(self.last_action, check))
         return check
 
-    def _remove_listener(self):
+    def _async_remove_listener(self):
         """Remove point in time listener, if any."""
-        if self._unsub_delay_listener:
-            self._unsub_delay_listener()
-            self._unsub_delay_listener = None
+        if self._async_unsub_delay_listener:
+            self._async_unsub_delay_listener()
+            self._async_unsub_delay_listener = None
 
     def _log(self, msg):
         """Logger helper."""
         if self.name is not None:
             msg = "Script {}: {}".format(self.name, msg)
 
         _LOGGER.info(msg)
+
+    def _trigger_change_listener(self):
+        """Trigger the change listener."""
+        if not self._change_listener:
+            return
+
+        self.hass.async_add_job(self._change_listener)
```

Comment:
`async_add_job` will schedule it on the event loop if it is a coroutine function (using `loop.create_task`) or will add it to the worker pool if it is not.

Still made `self._change_listener` a coroutine 👍